spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy

```diff
@@ -0,0 +1,211 @@
+private java.lang.Object foo
+
+private java.lang.Object $spock_initializeFields() {
+    foo = 'bar\'
+}
+
+@org.spockframework.runtime.model.FeatureMetadata(name = 'a feature', ordinal = 0, line = 3, blocks = [org.spockframework.runtime.model.BlockKind.SETUP[]], parameterNames = [])
+public void $spock_feature_0_0() {
+    java.lang.Object nothing = null
+    this.getSpecificationContext().getMockController().leaveScope()
+}'''
+  }
+
+
+  def "astToSourceFeatureBody can render everything"() {
+    when:
+    def result = compiler.astToSourceFeatureBody('''
+    given:
+    def nothing = null
+    ''', Show.all(), CompilePhase.INSTRUCTION_SELECTION)
+
+    then:
+    result == '''\
+package apackage
+
+import spock.lang.*
+
+@org.spockframework.runtime.model.SpecMetadata(filename = 'scriptXXXXX.groovy', line = 1)
+public class apackage.ASpec extends spock.lang.Specification {
+
+    @org.spockframework.runtime.model.FeatureMetadata(name = 'a feature', ordinal = 0, line = 1, blocks = [org.spockframework.runtime.model.BlockKind.SETUP[]], parameterNames = [])
+    public void $spock_feature_0_0() {
+        java.lang.Object nothing = null
+        this.getSpecificationContext().getMockController().leaveScope()
+    }
+
+}'''
+  }
+
+  def "full feature exercise"() {
+    when:
+    def result = compiler.astToSource('''
+    package apackage
+
+    import groovy.transform.EqualsAndHashCode
+
+    abstract class Foo implements Comparable {
+      List<String> x = new ArrayList<>()
+
+
+    }
+
+    @EqualsAndHashCode
+    class Bar extends Foo {
+      static final int[] ARR = [1, 2, 3] as int[]
+      static final String STR = 'str'
+      static final char CHR = ' '
+      static final int INT = 10
+      static int st
+      int order
+
+      Bar() {
+       println '42'
+      }
+
+      static {
+        st = 42
+      }
+
+      {
+        order = 1
+      }
+
+      void method() {
+        outer:
+        for (int i = 0; i < INT; i++) {
+          if(i % 2 == 0) {
+            print 'even'
+            continue outer
+          } else {
+            print 'odd'
+          }
+          break outer
+        }
+        for(String y : x) {
+          print(y)
+        }
+      }
+
+      @Override
+      int compareTo(Object o){
+        return order <=> (o as Bar).order
+      }
+    }
+
+    class Ext <T extends Serializable> {
+      def foo (List<? super T> consumer){}
+    }
+    ''')
+
+    then:
+    result == '''\
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+public abstract class apackage.Foo extends java.lang.Object implements java.lang.Comparable {
+
+    private java.util.List<String> x
+
+}
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+@groovy.transform.EqualsAndHashCode
+public class apackage.Bar extends apackage.Foo {
+
+    private static final [I ARR
```

Comment:
Can we render this properly as `int[]`?
And where is the initializer?

Reply:
That is something still to be seen, if you look at the comments in `spock.util.AstNodeToScriptVisitor#visitField`, we might choose to change that behavior.

> // do not print initial expression, as this is executed as part of the constructor, unless on static constant

as for the type that is handled in `spock.util.AstNodeToScriptVisitor#visitType` this might require a switch statement for every primitive array type, or do you have a better suggestion

Reply:
> // do not print initial expression, as this is executed as part of the constructor, unless on static constant

I cannot really follow the sense of that comment.
Why should where something is executed influence where it is in the script?
And even if it would make sense for the Groovy Console which I also dare to doubt, we want to check that the AST is like expected, so we should print it where it is in the AST.
Besides that it is simply swallowed completely, it is also not printed in the constructor but simply suppressed if it is not static final.
I would vote for always printing it in-place, that seems to be appropriate for our use-case imho.

> as for the type that is handled in spock.util.AstNodeToScriptVisitor#visitType this might require a switch statement for every primitive array type, or do you have a better suggestion

Maybe this is too naive and misses cases I'm not aware of, or it is simply that the class is 11 years old, but I'd say:
```groovy
  void visitType(ClassNode classNode) {
    print classNode.array ? classNode.toString(false) : classNode.name
    visitGenerics classNode?.genericsTypes
  }
```
The best would probably be we contribute those changes back upstream and see what the Groovy guys say and whether they see something else that is not working with the changes. :-)

Reply:
Or probably
```groovy
  void visitType(ClassNode classNode) {
    if (classNode.array) {
      print classNode.toString(false)
    } else {
      print classNode.name
      visitGenerics classNode?.genericsTypes
    }
  }
```

Reply:
I agree that we should print the field initializers.
The `visitType` needs to be improved anyway, since it does cause issues with generics.

Reply:
I tested 
```
  void visitType(ClassNode classNode) {
      print classNode.toString(false)
  }
```

and it works well in most of the cases
```

public class apackage.another.Ext<T extends java.io.Serializable, V extends java.lang.Cloneable> extends java.lang.Object {

    private static final int[] ARR = (int[]) [1, 2, 3]
    private T[] arr
    private java.util.List <V> lst = []

    public V foo(java.util.List <? super T> consumer) {
    }

    @apackage.another.Ann
    public boolean saveCompare(X a, X b) {
    }
}
```

it does loose the information in `saveCompare` although the `<X extends java.io.Serializable & java.lang.Comparable<T>>` should actually be in front of the `boolean` here. Maybe you can think of a good heuristic on when to print it.

Reply:
Part of the suggested changes, the method has generics information it is just not considered